components/core/src/clp/BoundedReader.hpp

```diff
@@ -0,0 +1,78 @@
+ *
+ * This is useful when the underlying input stream is divided into several logical segments and we
+ * want to prevent a reader for an earlier segment consuming any bytes from a later segment. In
+ * particular, reading part of a later segment may force the reader for that later segment to seek
+ * backwards, which can be either inefficient or impossible for certain kinds of input streams.
+ */
+class BoundedReader : public ReaderInterface {
+public:
+    // Constructor
+    explicit BoundedReader(ReaderInterface* reader, size_t bound)
+            : m_reader(reader),
+              m_bound(bound) {
+        if (nullptr == m_reader) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+        m_curr_pos = m_reader->get_pos();
+        if (m_curr_pos > m_bound) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+    }
+
+    // Methods implementing the ReaderInterface
+    /**
+     * Tries to get the current position of the read head in the underlying reader.
+     * @param pos Position of the read head in the underlying reader
+     * @return ErrorCode_Success on success
+     * @return ErrorCode_errno on failure
+     */
+    auto try_get_pos(size_t& pos) -> ErrorCode override { return m_reader->try_get_pos(pos); }
+
+    /**
+     * Tries to seek to the given position, limited by the checkpoint. If pos is past the checkpoint
+     * then this reader seeks up to the checkpoint in the underlying stream, then returns
+     * ErrorCode_EndOfFile on success or ErrorCode_errno otherwise.
+     * @param pos
+     * @return ErrorCode_Success on success
+     * @return ErrorCode_EndOfFile on EOF or if trying to seek beyond the checkpoint
+     * @return ErrorCode_errno on failure
+     */
+    auto try_seek_from_begin(size_t pos) -> ErrorCode override;
+
+    /**
+     * Tries to read up to a given number of bytes from the file, limited by the checkpoint. If the
+     * requested read would advance the read head beyond the checkpoint then the read size is
+     * adjusted such that the read head is advanced to the checkpoint. If a read is attempted when
+     * the read head is at the checkpoint then ErrorCode_EndOfFile is returned.
+     * @param buf
+     * @param num_bytes_to_read The number of bytes to try and read
+     * @param num_bytes_read The actual number of bytes read
+     * @return ErrorCode_errno on error
+     * @return ErrorCode_EndOfFile on EOF or trying to read after hitting checkpoint
+     * @return ErrorCode_Success on success
+     */
+    auto
+    try_read(char* buf, size_t num_bytes_to_read, size_t& num_bytes_read) -> ErrorCode override;
+
+    auto try_read_to_delimiter(char delim, bool keep_delimiter, bool append, std::string& str)
+            -> ErrorCode override {
+        return ErrorCode_Unsupported;
+    }
+
+private:
+    ReaderInterface* m_reader{nullptr};
```

Comment:
Using a pointer here doesn't enforce the given reader interface to have a lifetime >= the bounded reader. To be safe, shall we consider using a shared pointer instead?
What do u think? @kirkrodrigues

Reply:
I understand where you're coming from, but I decided not to use shared_ptr here for a few reasons.

1. The lifetime invariant is fairly simple to maintain in context, and any code that's using BoundedReader for its intended use will maintain the invariant automatically.
Specifically BoundedReader acts as a way to safely delegate ranges of an input stream to a series of different readers, which each relinquish (usually explicitly) the BoundedReader once they're finished with their section; the owner has to hold onto the underlying reader through all of these delegations so that it can correctly close the input stream at the end.

You would have to use BoundedReader in a somewhat contrived way outside of its intended use-case to violate the lifetime invariant, and anyone who does so will likely notice because their code will crash.

2. Forcing shared_ptr pollutes shared_ptr throughout the codebase even for very simple uses.

3. Forcing shared_ptr adds meaningful overhead in contexts where you may want to delegate thousands of small ranges (e.g. when reading tables from an archive).

4. Directly passing a pointer to another reader follows our existing convention for wrapping readers (e.g. our decompression classes taking raw FileReader*)